### PR TITLE
[JavaScript] Overhaul binding declarations. Feedback requested.

### PR DESCRIPTION
**Pull request by Thom1729**
Modern JavaScript offers destructuring assignment:

```js
const [foo, bar] = [1, 2];
console.log(foo); // 1
console.log(bar); // 1
```

Although this was not explicitly implemented in the syntax definition, it mostly worked in many common cases. I'm implementing it properly to work in all cases; then, I will implement destructuring for function parameters (#1102). This will involve a fair bit of new code as well as changes to existing code.

Before I continue, I'm looking for feedback on a couple of points:

- [Removing `readwrite` from variable scopes.](https://github.com/sublimehq/Packages/issues/1336)
- [Removing special handling of dollar signs.](https://github.com/sublimehq/Packages/issues/1337)

### CODE REVIEW

**Comment by Thom1729** [comment]
For the time being, I'll leave in both dollar identifiers and the `readwrite` scopes.

File: JavaScript/tests/syntax_test_js_bindings.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// SYNTAX TEST "Packages/JavaScript/JavaScript.sublime-syntax"
+
+const x = value;
+//    ^ meta.binding.name variable.other.readwrite
+
+
+const [ x, y, ...z, ] = value;
+//    ^^^^^^^^^^^ meta.binding.destructuring.sequence
```
Comment (wbond) [inline JavaScript/tests/syntax_test_js_bindings.js]: These assertions should probably continue for another 4 characters, right?

File: JavaScript/tests/syntax_test_js_bindings.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+//       ^ punctuation.separator.comma
+//         ^ meta.binding.name variable.other.readwrite
+//          ^ punctuation.separator.comma
+//            ^^^ keyword.operator.spread
+//               ^ meta.binding.name variable.other.readwrite
+//                ^ punctuation.separator.comma
+
+const [ x, [a, b], z] = value;
+//    ^^^^^^^^^^^^^^^ meta.binding.destructuring.sequence
+//         ^^^^^^ meta.binding.destructuring.sequence meta.binding.destructuring.sequence
```
Comment (wbond) [inline JavaScript/tests/syntax_test_js_bindings.js]: This should probably be `meta.binding.destructuring.sequence meta.binding.destructoring.sequence` to ensure it is doubled up, right?

File: JavaScript/tests/syntax_test_js_bindings.js (hunk 1 above)
Comment (Thom1729) [reply JavaScript/tests/syntax_test_js_bindings.js]: Yes.

File: JavaScript/tests/syntax_test_js_bindings.js (hunk 2 above)
Comment (Thom1729) [reply JavaScript/tests/syntax_test_js_bindings.js]: Is there a typo? I'm not seeing it.

Comment (wbond) [reply JavaScript/tests/syntax_test_js_bindings.js]: Sorry, it must have been wrapped on my screen last night and I didn't notice the second scope name.

**Pull request merged**